src/net.cpp

```diff
@@ -1312,46 +1296,153 @@ void CConnman::SocketHandler()
-            hSocketMax = std::max(hSocketMax, pnode->hSocket);
-            have_fds = true;
-
+            error_set.insert(pnode->hSocket);
             if (select_send) {
-                FD_SET(pnode->hSocket, &fdsetSend);
+                send_set.insert(pnode->hSocket);
                 continue;
             }
             if (select_recv) {
-                FD_SET(pnode->hSocket, &fdsetRecv);
+                recv_set.insert(pnode->hSocket);
             }
         }
     }
 
-    int nSelect = select(have_fds ? hSocketMax + 1 : 0,
-                         &fdsetRecv, &fdsetSend, &fdsetError, &timeout);
+    return !recv_set.empty() || !send_set.empty() || !error_set.empty();
+}
+
+#ifdef USE_POLL
+void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_set, std::set<SOCKET> &error_set)
+{
+    std::set<SOCKET> recv_select_set, send_select_set, error_select_set;
+    if (!GenerateSelectSet(recv_select_set, send_select_set, error_select_set)) {
+        interruptNet.sleep_for(std::chrono::milliseconds(SELECT_TIMEOUT_MILLISECONDS));
+        return;
+    }
+
+    std::vector<struct pollfd> pollfds;
+    pollfds.reserve(recv_select_set.size() + send_select_set.size() + error_select_set.size());
+
+    for (SOCKET hSocket : recv_select_set) {
+        struct pollfd pollfd = {};
+        pollfd.fd = hSocket;
+        pollfd.events = POLLIN;
+        pollfds.push_back(pollfd);
+    }
+
+    for (SOCKET hSocket : send_select_set) {
+        struct pollfd pollfd = {};
+        pollfd.fd = hSocket;
+        pollfd.events = POLLOUT;
+        pollfds.push_back(pollfd);
+    }
+
+    for (SOCKET hSocket : error_select_set) {
+        struct pollfd pollfd = {};
+        pollfd.fd = hSocket;
+        // These flags are ignored, but we set them for clarity
+        pollfd.events = POLLERR|POLLHUP;
+        pollfds.push_back(pollfd);
+    }
+
+    if(poll(pollfds.data(), pollfds.size(), 50) < 0)
+        LogPrint(BCLog::NET, "poll failure %d %s\n", errno, strerror(errno));
+
+    if (interruptNet)
+        return;
+
+    for (struct pollfd pollfd : pollfds) {
+        if (pollfd.revents & POLLIN)
+            recv_set.insert(pollfd.fd);
+        if (pollfd.revents & POLLOUT)
+            send_set.insert(pollfd.fd);
+        if (pollfd.revents & (POLLERR|POLLHUP))
+            error_set.insert(pollfd.fd);
+    }
+}
+#else
+void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_set, std::set<SOCKET> &error_set)
+{
+    std::set<SOCKET> recv_select_set, send_select_set, error_select_set;
+    if (!GenerateSelectSet(recv_select_set, send_select_set, error_select_set)) {
+        interruptNet.sleep_for(std::chrono::milliseconds(SELECT_TIMEOUT_MILLISECONDS));
+        return;
+    }
+
+    //
+    // Find which sockets have data to receive
+    //
+    struct timeval timeout;
+    timeout.tv_sec  = 0;
+    timeout.tv_usec = SELECT_TIMEOUT_MILLISECONDS * 1000; // frequency to poll pnode->vSend
```

Comment:
don't know how long the old code waited, but isn't 50 milliseconds a bit short? if there are no events, why not sleep as long as possible? can you please add a comment explaining the reasoning for this

Reply:
I carried over the existing value.

There are two reasons for this to be a short value:

- we cant modify the list of sockets while waiting, so the sleep time is potentially the delay before a new connection is serviced at all
- the disconnectnodes and inactivitychecks are time based and independent of socketevents

Reply:
i just checked and this is the value from satoshi, so im sure it was entirely arbitrarily chosen